jwst/datamodels/library.py

```diff
@@ -0,0 +1,160 @@
+        """
+        Compute a "group_id" without loading the file as a DataModel
+
+        This function will return the meta.group_id stored in the ASDF
+        extension (if it exists) or a group_id calculated from the
+        FITS headers.
+        """
+        # use astropy.io.fits directly to read header keywords
+        # avoiding the DataModel overhead
+        try:
+            with fits.open(filename) as ff:
+                if "ASDF" in ff:
+                    asdf_yaml = asdf.util.load_yaml(io.BytesIO(ff['ASDF'].data.tobytes()), tagged=True)
+                    if group_id := asdf_yaml.get('meta', {}).get('group_id'):
+                        return group_id
+                header = ff["PRIMARY"].header
+                program_number = header["PROGRAM"]
+                observation_number = header["OBSERVTN"]
+                visit_number = header["VISIT"]
+                visit_group = header["VISITGRP"]
+                sequence_id = header["SEQ_ID"]
+                activity_id = header["ACT_ID"]
+                exposure_number = header["EXPOSURE"]
+
+            return _attrs_to_group_id(
+                program_number,
+                observation_number,
+                visit_number,
+                visit_group,
+                sequence_id,
+                activity_id,
+                exposure_number,
+            )
+        except KeyError as e:
+            msg = f"Cannot find header keyword {e} in {filename}"
+            raise NoGroupID(msg) from e
+
+    def _model_to_group_id(self, model):
+        """
+        Compute a "group_id" from a model using the DataModel interface
+        """
+        if group_id := getattr(model.meta, "group_id", None):
+            return group_id
+        if hasattr(model.meta, "observation"):
+            return _attrs_to_group_id(
+                model.meta.observation.program_number,
+                model.meta.observation.observation_number,
+                model.meta.observation.visit_number,
+                model.meta.observation.visit_group,
+                model.meta.observation.sequence_id,
+                model.meta.observation.activity_id,
+                model.meta.observation.exposure_number,
+                )
+        raise NoGroupID(f"{model} missing group_id")
+
+    def _assign_member_to_model(self, model, member):
+        for attr in ("group_id", "tweakreg_catalog", "exptype"):
+            if attr in member:
+                setattr(model.meta, attr, member[attr])
+        if not hasattr(model.meta, "asn"):
+            model.meta["asn"] = {}
+
+        # if (model.meta.asn.table_name is None) and ("table_name" in self.asn.keys()): # do not clobber existing values
+        #     setattr(model.meta.asn, "table_name", self.asn["table_name"])
+
+        # if (model.meta.asn.pool_name is None) and ("asn_pool" in self.asn.keys()): # do not clobber existing values
+        #     setattr(model.meta.asn, "pool_name", self.asn["asn_pool"])
```

Comment:
yes, although it now occurs to me that after reverting this change, this might look the same as what is in `AbstractModelLibrary` so the whole function definition might be remove-able.  I will test this out